enhancements/authentication/login-logout-events.md

```diff
@@ -0,0 +1,164 @@
+
+2. **External OAuth identity providers**
+
+For login events external OAuth providers must invoke oauth-server callback handler to finalize authorization using [code grant flow](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1).
+
+Login success will be emitted upon successful code exchange with oauth-server in the [central OAuth callback handler](https://github.com/openshift/oauth-server/blob/822478f88514a80f053d9f65cd689d981b6ea4fd/pkg/oauth/external/handler.go#L149-L157). 
+
+Login failure however cannot be easily detected as there is no guarantee that external oauth providers will invoke callback in the case of a login failure. Hence, only best-effort login failure events can be emitted, if the external OAuth provider provides an `error` response in the [central `HandleRequest` handler](https://github.com/openshift/oauth-server/blob/513a8bb5b6cbd5e8faacbed523eb861aa29a674b/vendor/github.com/RangelReale/osincli/authorize.go#L84).
+
+3. **External OAuth identity providers used as challengers**
+
+External oauth identity providers used as challengers use a [central location](https://github.com/openshift/oauth-server/blob/822478f88514a80f053d9f65cd689d981b6ea4fd/pkg/oauth/external/handler.go#L118) where login and login failures audit events will be emitted.
+
+4. **Request Header identity provider**
+
+For this case an established trust to the external login proxy exists via means of client certificate trust. This implies that all authentication requests, naturally, have already passed successful authentication in the external 3rd party proxy.
+
+Thus, successful Login events can easily be constructed by creating audit events in the [central request authentication method](https://github.com/openshift/oauth-server/blob/4e63f0f350f28171edc4ae2553c7b852e65efbc6/pkg/authenticator/request/headerrequest/requestheader.go#L34).
+
+Login failure events, however never cause a request header authentication request to be generated in the 3rd party proxy, thus those events cannot be detected in oauth-server for this use case.
+
+### Logout events
+
+Logout events are not visible to oauth-server as the infrastructure in OpenShift does not have a central logout handler mechanism. Instead we have to rely on audit events against deletion of `oauthaccesstoken` resources.
+
+It is an open question how we can add information about the underlying user of the deleted `oauthaccesstoken` resource.
+
+### Event construction
+
+In all cases above a [Info](https://github.com/openshift/oauth-server/blob/961295a3151da3fff9daa5ea0ab8c0bf92a7d7e6/vendor/k8s.io/apiserver/pkg/authentication/user/user.go#L20) entity is available with user information which can be used to fill audit event data.
+
```

Comment:
@s-urbaniak  Just wanted to clairfy: for the logging of failed logins, will the output be:
1. a new event in audit logs
2. an entirely new log from apiserver
3. both
4. other
What logs will we have this event in basically?

Reply:
@EmilyM1 it is option 1. "a new event in audit logs"

Reply:
Ok, so they are independent. Oauth-server log will not go thru the kube-apiserver which currently outputs audit logs?
So no changes will be needed to kube-apiserver here to enable oauth-server to output audit logs?

Reply:
> Oauth-server log will not go thru the kube-apiserver which currently outputs audit logs?
correct

> So no changes will be needed to kube-apiserver here to enable oauth-server to output audit logs?
correct